### PR TITLE
Add possibility for a client to cancel a build…

### PR DESCRIPTION
**Pull request by vdemeester**
… by updating the status of a build.

It will mark the build as failed and clean any pods related to it.

Related to https://github.com/knative/build-pipeline/issues/272
I tried that on `knative/build` first (as it's simpler :upside_down_face:)

cc @abayer @ImJasonH 

Signed-off-by: Vincent Demeester <vdemeest@redhat.com>

### CODE REVIEW

**Comment by vdemeester** [comment]
@ImJasonH 

> 1.) Is it normally common practice for a client to be able to update k8s resources' status this way? I'm thinking of something like Job, which runs to completion, but which if you want to cancel it you basically just delete it. In this case, we want to stop the build's execution without also deleting any record of that resource existing.

I really am not sure… I'm trying to look around for that :sweat_smile: 

> 2.) Once this is done we should also make the same change for TaskRun, and then PipelineRun, which should cancel (but not delete) its created TaskRuns. We might even just want to go straight to TaskRun and drop this PR. I'll leave that decision to up to you.

Yes, I did this here as it was simpler/quicker but I intend to follow-up on `build-pipeline`. I'm also fine doing it directly on `build-pipeline`, I just thought if we want it in `build-pipeline` we *may* want it there too :angel:

**Comment by shashwathi** [comment]
@vdemeester Great PR

If build is completed then what is the point updating status to `Cancelled`?
Build is completed so all associated resources are deleted.
If build is running then `cluster` name would be set in build status. 
So I don't understand why there is a API call to get pod resource? and why not just call the delete API directly?

**Comment by vdemeester** [comment]
> @vdemeester Great PR
> 
> If build is completed then what is the point updating status to `Cancelled`?
> Build is completed so all associated resources are deleted.

So, if you want to cancel the build (currently using the spec), then you want to mark it as `Cancelled` as soon as possible and stop the execution, thus deleting the pod real quick :angel:, same way we do on build timeout :upside_down_face:. (making me think there could be a small refactor to share code there).

> If build is running then `cluster` name would be set in build status.
> So I don't understand why there is a API call to get pod resource? and why not just call the delete API directly?

Indeed, there is no need.

**Comment by abayer** [comment]
At Kubecon, @bobcatfish told me (in the context of https://github.com/knative/build-pipeline/pull/355) that I shouldn't be making changes to the status of one CRD while reconciling another one, so I jumped through some hoops to handle having the `PipelineRun`'s timeout apply to its `TaskRun`s. Not sure if that is a general thing where nothing should be modifying a CRD's status but its reconciler or if it's just that one CRD's reconciler shouldn't mess with another CRD's status.

**Comment by shashwathi** [comment]
> So, if you want to cancel the build (currently using the spec), then you want to mark it as Cancelled as soon as possible and stop the execution, thus deleting the pod real quick 👼

I agree with this completely. I do not think you understood my question so let me rephrase
If build is completed (pods are not present and only build object exists), then what is the point of marking build as "Cancelled"?

I am thinking of this like a cycle.
build starts -> pod creation -> pod finished and deleted -> build completed. 

If user marks build as "Cancelled" after completion, I do not see how that has any effect on underlying resource(pod). I would expect in this case for status to be not changed at all. If build has finished then status should not be changed(IMO).

> same way we do on build timeout 🙃. (making me think there could be a small refactor to share code there).

Controller checks if build is still running and then cancels the build if timeout has passed. It doesn't cancel if build has completed.

**Comment by vdemeester** [comment]
> I agree with this completely. I do not think you understood my question so let me rephrase
> If build is completed (pods are not present and only build object exists), then what is the point of marking build as "Cancelled"?
> 
> I am thinking of this like a cycle.
> build starts -> pod creation -> pod finished and deleted -> build completed.
> 
> If user marks build as "Cancelled" after completion, I do not see how that has any effect on underlying resource(pod). I would expect in this case for status to be not changed at all. If build has finished then status should not be changed(IMO).

Indeed there is no point, and if a user marks the build as `Cancelled` after completion, nothing should happen indeed. It should already be the case (https://github.com/knative/build/pull/510/files#diff-f2d9d6d36e3340295720d4f3e028cb14R152). If the build is completed (successfully or not), it won't ever go into the "cancel" block :angel: 

> Controller checks if build is still running and then cancels the build if timeout has passed. It doesn't cancel if build has already completed.

Yes, this is what happens.

File: test/e2e/simple_test.go [hunk 1]
```diff
@@ -605,3 +605,47 @@ func TestSimpleBuildWithHybridSources(t *testing.T) {
...
+				Args:  []string{"sleep", "5000"}, // fails.
+			}},
+		},
+	}); err != nil {
+		t.Fatalf("Error creating build: %v", err)
+	}
+
+	// Wait for a little while
+	// FIXME(vdemeester) I would prefer something less flaky-prone
+	time.Sleep(20 * time.Second)
```
Comment (shashwathi) [inline test/e2e/simple_test.go]: Probably consider adding polling function to check build status is "Running" for 10sec(? or more). That is better than wait solution IMO. WDYT @vdemeester ?

File: test/e2e/simple_test.go [hunk 2]
```diff
@@ -605,3 +605,47 @@ func TestSimpleBuildWithHybridSources(t *testing.T) {
...
+	if err != nil {
+		t.Fatalf("Error getting build %s: %v", buildName, err)
+	}
+	b.Spec.Status = v1alpha1.BuildSpecStatusCancelled
+	if _, err := clients.buildClient.builds.Update(b); err != nil {
+		t.Fatalf("Error update build status for %s: %v", buildName, err)
+	}
+
+	if _, err := clients.buildClient.watchBuild(buildName); err == nil {
+		t.Fatalf("watchBuild did not return expected `cancelled` error")
```
Comment (shashwathi) [inline test/e2e/simple_test.go]: `watchBuild` function returns `build` object so you could consider verifying that build status is updated with expected reason("BuildCancelled") here.

File: pkg/reconciler/build/build_test.go [hunk 3]
```diff
@@ -561,3 +561,62 @@ func TestTimeoutFlow(t *testing.T) {
...
+		t.Fatalf("build status did not specify podName: %v", b.Status.Cluster)
+	}
+	b.Spec.Status = v1alpha1.BuildSpecStatusCancelled
+
+	f.updateIndex(i, b)
+	if err := r.Reconcile(ctx, getKey(b, t)); err != nil {
+		t.Errorf("error syncing build: %v", err)
+	}
+
+	// Check that the build has the expected timeout status.
```
Comment (shashwathi) [inline pkg/reconciler/build/build_test.go]: typo: I think you meant cancelled status here

File: pkg/reconciler/build/build.go [hunk 4]
```diff
@@ -153,6 +153,36 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
...
+			Message: fmt.Sprintf("Build %q was cancelled", build.Name),
+		})
+		if err := c.updateStatus(build); err != nil {
+			return err
+		}
+		if build.Status.Cluster == nil {
+			logger.Warnf("build %q has no pod running yet", build.Name)
+			return nil
+		}
+		p, err := c.kubeclientset.CoreV1().Pods(build.Namespace).Get(build.Status.Cluster.PodName, metav1.GetOptions{})
```
Comment (shashwathi) [inline pkg/reconciler/build/build.go]: I think I already mentioned this(?) in previous comment about reusing `build.Status.Cluster.PodName` for Delete API directly. Sorry to repeat it again

File: test/e2e/simple_test.go (hunk 1 above)
Comment (vdemeester) [reply test/e2e/simple_test.go]: @shashwathi yeah, that's why the FIXME :angel: not sure why I didn't use the polling functions here :sweat_smile: will update :wink:

File: pkg/reconciler/build/build.go (hunk 4 above)
Comment (vdemeester) [reply pkg/reconciler/build/build.go]: oh right, didn't update that :sweat:

**Comment by shashwathi** [comment]
I noticed @ImJasonH had some comment earlier in thread so I am adding `hold` for him to add his review.

**Comment by ImJasonH** [comment]
/lgtm
/approve
/hold cancel

Oh my goodness I didn't realize this was blocking on me, so so sorry to keep this held up so long. Thanks for this change! :heart:

**Comment by vdemeester** [comment]
ah.. I may need to rebase and fix 😅

**Comment by vdemeester** [comment]
Hum… the coverage build is a bit a pain :sweat_smile: as the threshold is higher than the current coverage on that file, any change trigger a failure… :sweat:

**Comment by vdemeester** [comment]
| /test pull-knative-build-integration-tests
-- | --
